[feat] 구글 로그인/로그아웃,  리다이렉트 구현 (#16)

nakyeonko3 opened the pull request:
# 🚀 Pull Request Proposal

**[Please briefly describe the work done]**

## 📋 Work Details

### 로그인/로그아웃 기능 구현
- 구글/깃허브 로그인/로그아웃 구현
- 보호 라우팅을 이용해서 로그인 리다이렉트 구현
- 로그인시 홈으로 이동, 로그아웃시 로그인 페이지로 이동 구현
- 로그인 하지 않고 다른 페이지 이동을 막음.


### 로딩 페이지 및 글로벌 스타일 수정
-  refactor: 로딩페이지 추가 및 로그인 에러 핸들링 함수 추가   
- 로딩페이지 추가, src/components/LoadingPage.tsx, loading.tsx를 이용해서 로딩페이지를 구현
- 로그인 페이지에서 로딩페이지를 이용해서 로딩 중임을 표시하도록 수정
- 라우팅 페이지 로그인 페이지에서 로딩페이지를 이용해서 로딩 중임을 표시하도록 수정, src/routes/PublicRoute.tsx, src/routes/ProtectedRoute.tsx
- theme.ts에 primary 색상을 #ff654e로 변경
![firefox_20240924_13h49m_1904](https://github.com/user-attachments/assets/e3162d2c-2bca-4f66-b4ea-d40afaff54be)

### 로그인 에러 헨들링 관련 로직 수정
- 로그인 에러 핸들링 함수 추가, src/util/authErrorHandling.ts, firebaseAuth.ts에서 발생하는 에러를 핸들링하는 함수 추가
- 같은 이메일로 가입한 계정이 있을 경우 구글로그인이나 깃허브 로그인 중 하나만 하도록 수정함.


### .env 파일 수정
.env에 firebase auth관련 환경변수 값들이 추가되었습니다.
카톡으로 보내드리겠습니다.


## 🔧 Changes Summary

Summarize the key changes made.

## 📸 Screenshots (Optional)

You can attach screenshots demonstrating the modified screens or features.

## 📄 Additional Information

If you have any additional information or special requests, please include them here.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **새로운 기능**
	- Google 및 GitHub 로그인을 위한 새로운 인증 메서드 추가.
	- 설정 페이지에 로그아웃 버튼 추가.
	- 공용 및 보호된 경로를 관리하는 `PublicRoute` 및 `ProtectedRoute` 컴포넌트 추가.
	- 로딩 인디케이터를 표시하는 새로운 `LoadingPage` 컴포넌트 추가.
	- 버튼 컴포넌트에 클릭 이벤트를 처리할 수 있는 선택적 `onClick` 속성 추가.
	- 로그인 페이지에 Google 및 GitHub 로그인 기능 통합.
	- 새로운 `authErrorHandling` 유틸리티 함수 추가.

- **버그 수정**
	- Google 로그인 실패 시 잘못된 오류 메시지 수정.

- **스타일**
	- 테마 색상 정의 수정.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

nakyeonko3 committed:
feat 로그인/로그아웃 리다이렉트 구현 (#16) (#16)

badaclockA commented on src/api/firebaseAuth.ts:
```diff
@@ -44,3 +44,16 @@ export const signOutWithGithub = async () => {
     });
 };
+
+export const signInWithGoogle = async () => {
+  const provider = new GoogleAuthProvider();
+  await signInWithPopup(auth, provider)
+    .then((result) => {
+      const user = result.user;
+      console.log(`user:`, user);
+    })
+    .catch((error) => {
+      // TODO: 로그인 실패시 에러 처리 로직을 작성해야함
+      console.error('GitHub 로그인 실패:', error.message);
+    });
+};
```
signInWithPopup.then을 해야지 팝업을 잘못 클릭했을 때 오류 처리를 할 수 있는 것 같다. 아닌가?

badaclockA replied:
throw error 와 throw new Error("에러메시지") 와 차이가 있나나

badaclockA commented on src/pages/Login.tsx:
```diff
@@ -22,6 +26,7 @@ const Login = () => {
             fontColor="#fff"
             image={githubLogo}
             text="Log in with Github"
+            onClick={() => githubLogin()}
```
useGithubSignIn에 이미 오류 처리 로직이 들어있는데 Login.tsx에도 넣어야 하나?
```
  const signIn = async () => {
    try {
      await signInWithGithub();
      navigate('/');
    } catch (error) {
      console.error('깃허브 아이디를 이용해 로그인하는데 실패했습니다.', error);
      //TODO: 로그인 실패시 사용자에게 알림을 주는 코드를 작성해야함
    }
  };

```

nakyeonko3 commented on src/routes/ProtectedRoute.tsx:
```diff
@@ -0,0 +1,14 @@
+export const ProtectedRoute: React.FC = () => {
+  const { user, loading } = useUserAuthentication();
+
+  if (loading) {
+    return <div>Loading...</div>;
+  }
```
한번 
lucid 아이콘을 써도다좋다

nakyeonko3 committed:
refactor: 로딩페이지 추가 및 로그인 에러 핸들링 함수 추가

- 로딩페이지 추가, src/components/LoadingPage.tsx, loading.tsx를 이용해서 로딩페이지를 구현
- 로그인 페이지에서 로딩페이지를 이용해서 로딩 중임을 표시하도록 수정
- 로그인 에러 핸들링 함수 추가, src/util/authErrorHandling.ts, firebaseAuth.ts에서 발생하는 에러를 핸들링하는 함수 추가
- 라우팅 페이지 로그인 페이지에서 로딩페이지를 이용해서 로딩 중임을 표시하도록 수정, src/routes/PublicRoute.tsx, src/routes/ProtectedRoute.tsx
- theme.ts에 primary 색상을 #ff654e로 변경

nakyeonko3 committed:
fix:firebaseAuth.ts 수정(#16)

- 로그인 실패시 throw error를 통해 호출자가 에러를 감지하도록함.

badaclockA commented on src/api/firebaseAuth.ts:
```diff
@@ -2,45 +1,45 @@
-      console.error('로그아웃 실패', error);
-    });
+export const signInWithGoogle = async (): Promise<void> => {
+  const provider = new GoogleAuthProvider();
+  try {
+    const result = await signInWithPopup(auth, provider);
+    const credential = GoogleAuthProvider.credentialFromResult(result);
+    if (credential) {
+      // This gives you a Google Access Token. You can use it to access the Google API.
+      // const token = credential.accessToken;
+      // console.log('token:', token);
+    }
+  } catch (error) {
+    console.error('구글 아이디를 이용해 로그인하는데 실패했습니다.', error);
+  }
```
호출자가 에러를 감지한다는 뜻이 뭐지지

nakyeonko3 committed:
chore: byul종속성을  1.4.1로 변경(#16)

theconomicat reviewed (approved):
Thank you
